airflow/www/extensions/init_wsgi_middlewares.py

```diff
@@ -37,8 +38,11 @@ def _root_app(env: WSGIEnvironment, resp: StartResponse) -> Iterable[bytes]:
 
 def init_wsgi_middleware(flask_app: Flask) -> None:
     """Handle X-Forwarded-* headers and base_url support."""
+    webserver_base_url = conf.get_mandatory_value("webserver", "BASE_URL", fallback="")
+    if webserver_base_url.endswith("/"):
+        raise AirflowConfigException("webserver.base_url conf cannot have a trailing slash.")
```

Comment:
I am wondering here that why can the base URL not have a trailing '/'?

Reply:
This is due to how web servers rewrite URLs when you deploy a webapp under a prefix. Say you deploy under `https://mydomain/airflow`. When you visit `https://mydomain/airflow/home` in the browser (as an example; any client applies), the proxy server (e.g. NGINX) would strip the prefix, and forward the rest of the URL to the app, so in Python code (e.g. Airflow) we can just handle `/home` without needing to consider what prefix the entirety of Airflow is deployed under.

Now you may say, hey, why can’t either the proxy server, the gateway protocol (WSGI in Python’s case), or the web framework (Flask for Airflow’s case) be smarter and just strip or add the slash in between as appriproate? And you would be right! Some of them actually can do this (NGINX has `merge_slashes`, for example), but not everyone does since the prefix _technically_ should never have a trailing slash (the slash is _technically_ significant in a URL), and that extra check is not free. So it’s better to avoid fighting the tools and just stick to the technically correct configuration.

Reply:
okay, I am coming from [here](https://github.com/apache/airflow/issues/32996#issuecomment-1659952828) on how the python urllib.parse.urljoin handles joining base URL and relative URLs.

Wouldn't `https://my.astronomer.run/path` qualify as a valid base URL? But looks like the `urljoin` just ignores the path when it does not end with a trailing `/`. If it was `https://my.astronomer.run/path/` `urljoin` does not strip the path

Reply:
I think we can handle more cases if users need it. This PR was a quick fix to prevent circular redirections.

I’m sure we can make this work with a trailing slash but requires a bit more work.

Reply:
(Sorry, forgot to reply.)

Python’s `urljoin` implements the logic in `<a href="...">` tags; I don’t recall the correct term for this, but it is the URL resolution logic when you click on links in a browser. URLs have two kinds, folder and file (not the right terms, but easier to understand). The URL would not have a trailing slash if you are in a file view; the trailing slash indicates a folder view. When you’re in a file, a relative path like `foo` indicates another file in the same folder. When you’re in a folder view, however, `foo` means the foo entry _in_ the folder. This means the resolution logic would change depending on whether the URL has a trailing slash or not.

Note that Python’s local path joining logic (both pathlib and os.path) is different and does not consider the trailing slash; it instead matches the common logic used to join format in shell scripts.

But the prefix is an entirely different thing, and is intended to only be joined with simple string concatenation and skips all the folder/file logic because it is not semantically possible to support combining the prefix with absolute paths and relative paths. The only proper way to prepend a prefix is `prefix + path` (or other string concatenation methods such as `f"{prefix}{path}"`, of course).

Reply:
understood, thank you @uranusjr for the detailed explanation.